util/coolutils/uc.1
.\"
.\"
.\" uc.man - uc manpage
.\" Copyright (C) 2010 Lenik (谢继雷)
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH UC 1
.SH NAME
uc \- Translate to specific cases
.SH SYNOPSIS
.B uc
[OPTIONS] [ FILES | STRING-ARGS ]
.PP

.B lc
[OPTIONS] [ FILES | STRING-ARGS ]
.PP

.B to-caption
[OPTIONS] [ FILES | STRING-ARGS ]
.PP

.B to-dash
[OPTIONS] [ FILES | STRING-ARGS ]
.PP

.B to-camel
[OPTIONS] [ FILES | STRING-ARGS ]
.PP

.B to-mech
[OPTIONS] [ FILES | STRING-ARGS ]

.SH DESCRIPTION
Translate to specific cases

.SH OPTIONS

.SH EXAMPLES

.B
uc -e hello, World
.PP
Convert arguments to upper case.

.B
uc -l Hello, WORLD
.PP
Convert arguments to lower case. (Same as `lc Hello, WORLD')

.SH ENVIRONMENT

.SH REPORTING BUGS
Please report any bugs, problems, queries, experiences, etc. directly to the author.

.SH AUTHORS
Lenik <lenik at bodz.net>
.br


ChangeLog
uni-2.14.11 (Lenik, 2013-09-09)
	* Dispersed Lapiota/j into Jazz uni-project.
	* shelj package created by newdeb
	* visualcss package created by newdeb
	* jsrcutils package created by newdeb
	* winedit package created by newdeb
	* coolclip: This package is reserved for adding extra format supports in clipboard.

uni-2.14.10 (Lenik, 2013-09-07)
	* snmcc: Renamed from lapiota-snmcc.
	* pgputils: Separated from lapiota-cli.
	* pgputils package created by newdeb
	* lapiotaj: Moved lapiota-api to libuni-java, and XMLFlat/Edit to xmlproc.
	* libuni-java package created by newdeb
	* xmlproc package created by newdeb
	* regf2t: Changed package name to ..fmt.regf2t.
	* Initialized Maven multi-module project on uni.
	* Refactor the whole directory structures. (...)
	* vcscmds: Print hline for each dir in vcsdirs.
	* bash-shlib: Added term to print hline.

uni-2.14.9 (Lenik, 2013-09-06)
	* regf2t: Added data structs. (Ref: reglookup, samba4)
	* regf2t package created by newdeb
	* lapiotaj: Refactor ..sio to io.*.

uni-2.14.8 (Lenik, 2013-09-05)
	* lapiotaj: Memory was renamed to IMemory.

uni-2.14.7 (Lenik, 2013-09-04)
	* wine-enh: The dependency winetricks is optional.

uni-2.14.6 (Lenik, 2013-09-03)
	* diskutils: Added read/writembr, and merged mkhdd/partoffset.
	* diskutils package created by newdeb

uni-2.14.5 (Lenik, 2013-08-30)
	* echo, lapiotaj: New Jazz project/module metadata design.

uni-2.14.4 (Lenik, 2013-08-24)
	* x-tools-local: Setup PATH env for crosstool-ng.
	* x-tools-local package created by newdeb
	* users-pref-xjl: Shift 70-groups to 700+.
	* cdu-bodz-svn: Upgraded site conf for apache-2.4.
	* cdu-bodz-trac: Upgraded site conf for apache-2.4.
	* cdu-bodz-m2: Upgraded site conf for apache-2.4.
	* cls-a2sites: Upgraded site conf for apache-2.4.
	* dlocal: Upgraded site conf for apache-2.4.

uni-2.14.3 (Lenik, 2013-08-23)
	* imgauto: Added option to jpeg-ff: --odd.
	* lapiotaj: bas-memory was moved to bas-data.
	* lapiotaj: Organized imports.
	* lapiota-cli: _help() was renamed to showHelpPage().

uni-2.14.2 (Lenik, 2013-08-22)
	* lapiotaj: INegotiation was refactored to IOptions.
	* lapiotaj: "trait" was changed to "mf".

uni-2.14.1 (Lenik, 2013-08-13)
	* coolapt: Added apt-list-dep, to mark dependencies as auto.

uni-2.14.0 (Lenik, 2013-08-08)
	* cdu-bodz-uni package created by newdeb
	* README: Address changes.
	* imgauto: Added jpeg-ff to generate flip-flop repeated images.
	* imgauto package created by newdeb

uni-2.13.1 (Lenik, 2013-07-16)
	* virtualbox-enh: Added mkhdd.

uni-2.13.0 (Lenik, 2013-07-03)
	* keep empty dirs
	* coollog: Added chlog (-g, -G to use git) to trace file changes.

uni-2.12.9 (Lenik, 2013-05-30)
	* vim-enh: Reload /etc/exports by `exportfs -ra`.
	* wine-enh: Added exe-runner to run .exe files with options. (Using binfmts)
	* rrdtool-enh: Install test/demo.rrd to /usr/share/rrdtool-enh/samples.
	* git-enh: Added alias gmt = git mergetool --tool=meld.
	* textcodecs: Added uri-decode/encode from lenik/bin.
	* textcodecs package created by newdeb
	* coolpdf: Added pdfcrop (from 3rdparty) and pdfrepair.
	* coolpdf package created by newdeb
	* astah-local: Added astah launcher.
	* astah-local package created by newdeb

uni-2.12.8 (Lenik, 2013-05-10)
	* coolsql: Added missing --console option.

uni-2.12.7 (Lenik, 2013-04-25)
	* mjpl: Initial.
	* bash-enh: Renamed env() to dv(), and envchk.sh to envop.sh.

uni-2.12.6 (Lenik, 2013-04-24)
	* 2utf8: Treat chardet/gb2312 as gbk. And use iconv -c option.

uni-2.12.5 (Lenik, 2013-04-23)
	* 7z-enh: Added unzip-utf8.

uni-2.12.4 (Lenik, 2013-04-20)
	* 7z-enh package created by newdeb

uni-2.12.3 (Lenik, 2013-04-17)
	* mate-pref-xjl: Changed "<Super>equal" to "<Mod4>equal".

uni-2.12.2 (Lenik, 2013-04-08)
	* mate-pref-xjl: Migrated from the mateconf backend to the dconf backend.

uni-2.12.1 (Lenik, 2013-04-03)
	* mjpl package created by newdeb
	* isz-tool: Packaging for Debian.
	* coolmaven: Renamed PROJECT-LIST to projects, MODULES to modules.tab, etc.

uni-2.12.0 (Lenik, 2013-04-02)
	* rsync-rrc: Merged rsync-enh/ r-pull/r-push to rrc.

uni-2.11.8 (Lenik, 2013-03-31)
	* svn-enh: Added sudosvn which try to use the private key of login user.
	* svn-enh: Added /etc/subversion/cls/conf for shared authz/passwd.
	* cdu-bodz-m2-settings: Added support for both maven2/maven3.

uni-2.11.7 (Lenik, 2013-03-29)
	* coolsql: Initial release with pg(postgresql) driver.
	* coolsql package created by newdeb

uni-2.11.6 (Lenik, 2013-03-15)
	* usecode: Added dependency to tty-fx for ⚝ input.
	* tty-fx: Renamed coolansi to tty-fx, and added stty ⚝ aliases.

uni-2.11.5 (Lenik, 2013-03-14)
	* lapiota/j: Utilized $project.version.
	* lapiota-cli: Replaced JavaLauncher+ to Shell.Main.

uni-2.11.4 (Lenik, 2013-02-19)
	* coolmaven: Added --max-permsize option to jetty.

uni-2.11.3 (Lenik, 2013-02-08)
	* bash-enh: Changed df alias to "df -T -B 1048576".
	* rsync-enh: Added r-push, r-pull.
	* rsync-enh package created by newdeb

uni-2.11.2 (Lenik, 2013-02-07)
	* echo-server: Configure server in init() rathern then ctor, so one can do more configs.

uni-2.11.1 (Lenik, 2013-02-05)
	* android-sdk-local: Add android-sdk/tools to PATH.

uni-2.11.0 (Lenik, 2013-02-01)
	* android-sdk-local package created by newdeb
	* x509utils: Changed default expire date to 50 years, and added pkcs12jks.

uni-2.10.25 (Lenik, 2013-01-31)
	* z-latex: Using CJK numbers in chapter names.
	* pgadmin3-pref-xjl: Added [ctlSQLBox] patch installer.
	* pgadmin3-pref-xjl package created by newdeb

uni-2.10.24 (Lenik, 2013-01-28)
	* archiva: Added activiti related repos.

uni-2.10.23 (Lenik, 2013-01-07)
	* echo-server: Moved http-servlet base types to bas-http.

uni-2.10.22 (Lenik, 2013-01-06)
	* lapiotaj: Refactor/ref: DefaultClassLoader -> ClassLoaders.runtimeClassLoader.

uni-2.10.21 (Lenik, 2013-01-05)
	* lapiotaj: Refactor/ref: StringEscape.escapeJava.

uni-2.10.20 (Lenik, 2013-01-04)
	* coolmaven: Added mjava -e option, and bash-completion support for search main classes.

uni/xjl-coolutil/uc.c
#include "config.h"

#include <unistd.h>
#include <stdio.h>
#include <string.h>

#define UPPER   0                       /* ALL UPPER CASE */
#define LOWER   1                       /* ALL LOWER CASE */
#define CAPTION 2                       /* In Caption Case */
#define DASH    3                       /* in dash-line */
#define CAMEL   4                       /* in mixedCase */
#define MECH    5                       /* In CaptionCase */

#define VOID    EOF

char *program;
int opt_echo = 0;
int opt_case = UPPER;
int opt_dash = '-';

const char *case_help[] = {
    "UPPER CASE",
    "lower case",
    "Caption Case",
    "dash-line case",
    "camelCase",
    "MechnicalCase",
};

typedef int (*TRFUNC)(int c);

int tr_upper(int c) {
    return toupper(c);
}

int tr_lower(int c) {
    return tolower(c);
}

int tr_caption(int c) {
    static int word_off = 0;
    if (isalpha(c)) {
        if (word_off)
            c = tolower(c);
        else
            c = toupper(c);
        word_off++;
    } else
        word_off = 0;
    return c;
}

int tr_dash(int c) {
    static int word_off = 0;
    if (isalpha(c)) {
        if (word_off)
            if (isupper(c)) {
                putchar(opt_dash);
                c = tolower(c);
            }
        word_off++;
    } else
        word_off = 0;
    return c;
}

int tr_camel(int c) {
    static int last = 0;
    int c0 = c;
    if (c == opt_dash)                  /* [-] */
        c = VOID;
    else if (last == opt_dash) {         /* -c */
        // printf("last=%c, c=%c\n", last, c);
        if (isalpha(c))
            c = toupper(c);             /* _alpha => Alpha */
        else
            putchar(opt_dash);          /* restore [-] */
    }
    last = c0;
    return c;
}

/* TODO - not implemented. */
int tr_mech(int c) {
    static int word_off = 0;
    // static char last = 0;
    if (isalpha(c)) {
        if (word_off)
            c = tolower(c);
        else
            c = toupper(c);
        word_off++;
    } else
        word_off = 0;
    return c;
}

void show_version() {
    printf("%s: translate text to %s\n", program, case_help[opt_case]);
    printf("Version 0.1,  Written by Lenik,  Aug 2010\n");
}

void show_help() {
    show_version();
    printf("\n");
    printf("Syntax: \n");
    printf("    -e, --echo       translate the arguments itself\n");
    printf("    -u, --upper      to UPPER CASE\n");
    printf("    -l, --lower      to lower case \n");
    printf("    -c, --caption    to Caption Case\n");
    printf("    -d, --dash       to dash-line case\n");
    printf("    -j, --camel      to camelCase\n");
    printf("    -m, --mech       to CaptionCase case\n");
    printf("    -h, --help       show this help page\n");
    printf("        --version    show version info\n");
}

int main(int argc, char **argv) {

    int c;
    int i;
    TRFUNC trf = NULL;

    program = argv[0];
    --argc; ++argv;

    char *program_base = strrchr(program, '/');
    if (! program_base)
        program_base = program;

    if (! strcmp(program_base, "uc"))
        opt_case = UPPER;
    else if (! strcmp(program_base, "lc"))
        opt_case = LOWER;
    else if (! strcmp(program_base, "to-caption"))
        opt_case = CAPTION;
    else if (! strcmp(program_base, "to-dash")) /* see: bsdmainutils:ul */
        opt_case = DASH;
    else if (! strcmp(program_base, "to-camel"))
        opt_case = CAMEL;
    else if (! strcmp(program_base, "to-mech"))
        opt_case = MECH;

    while (argc) {
        const char *arg = *argv;
        if (*arg == '-') {
            // printf("opt: %s\n", arg);
            argc--;
            argv++;
            arg++;
            if (*arg == '-') {          /* --? */
                arg++;
                if (! *arg)             /* -- */
                    break;
                if (! strcmp("echo", arg))
                    opt_echo = 1;
                else if (! strcmp("upper", arg))
                    opt_case = UPPER;
                else if (! strcmp("lower", arg))
                    opt_case = LOWER;
                else if (! strcmp("caption", arg))
                    opt_case = CAPTION;
                else if (! strcmp("dash", arg))
                    opt_case = DASH;
                else if (! strcmp("camel", arg))
                    opt_case = CAMEL;
                else if (! strcmp("mech", arg))
                    opt_case = MECH;
                else if (! strcmp("help", arg)) {
                    show_help();
                    return 0;
                }
                else if (! strcmp("version", arg)) {
                    show_version();
                    return 0;
                }
                else {
                    fprintf(stderr, "Invalid option: %s", arg);
                    return 1;
                }
            } // --
            else {
                while (c = *arg++) {
                    switch (c) {
                    case 'e':
                        opt_echo = 1; break;
                    case 'u':
                        opt_case = UPPER; break;
                    case 'l':
                        opt_case = LOWER; break;
                    case 'c':
                        opt_case = CAPTION; break;
                    case 'd':
                        opt_case = DASH; break;
                    case 'j':
                        opt_case = CAMEL; break;
                    case 'm':
                        opt_case = MECH; break;
                    case 'h':
                        show_help();
                        return 0;
                        break;
                    default:
                        fprintf(stderr, "Invalid option: %c\n", c);
                        return 1;
                    }
                }
            } // -x
        } // -
        else
            break;
    } // while argc

    switch (opt_case) {
    case UPPER:
        trf = tr_upper; break;
    case LOWER:
        trf = tr_lower; break;
    case CAPTION:
        trf = tr_caption; break;
    case DASH:
        trf = tr_dash; break;
    case CAMEL:
        trf = tr_camel; break;
    case MECH:
        trf = tr_mech; break;
    }

    if (opt_echo) {
        for (i = 0; i < argc; i++) {
            const char *s = *argv++;
            // printf("arg: %s\n", s);
            if (i) putchar(trf(' '));
            while (c = *s++) {
                c = trf(c);
                if (c != VOID)
                    putchar(c);
            }
        }
        printf("\n");
        return 0;
    }

    /* no filename or '-' is specified */
    if (argc == 0 ||
        (argc == 1 && ! strcmp(argv[0], "-"))) {
        while ((c = getchar()) != EOF) {
            c = trf(c);
            if (c != VOID)
                putchar(c);
        }
        return 0;
    }

    int file_index;
    for (file_index = 0; i < argc; i++) {
        char *filename = argv[file_index];
        // printf("file: %s\n", filename);
        FILE *f = fopen(filename, "rb");
        if (!f) {
            fprintf(stderr, "Failed to open file %s: ", filename);
            perror("");
            return 1;
        }
        while ((c = fgetc(f)) != EOF) {
            c = trf(c);
            if (c != VOID)
                putchar(c);
        }
        fclose(f);
    }
    return 0;
}
